lib/when_committed.rb

```diff
@@ -2,37 +2,41 @@
 
 module WhenCommitted
   module ActiveRecord
-    def self.included(base)
-      base.after_commit :run_when_committed_callbacks
-      base.after_rollback :clear_when_committed_callbacks
-    end
-
-    def when_committed(&block)
-      when_committed_callbacks << block
-    end
-
-    def when_committed!(&block)
-      if self.class.connection.open_transactions > 0
-        when_committed(&block)
+    def when_committed(run_now_if_no_transaction: false, &block)
+      if self.class.connection.current_transaction.open?
+        cb = CallbackRecord.new(&block)
+        self.class.connection.current_transaction.add_record(cb)
       else
-        block.call
+        if run_now_if_no_transaction
+          block.call
+        else
+          raise RequiresTransactionError
+        end
       end
     end
+  end
 
-    private
+  # Adheres to the "record" duck type expected by the `add_record` method on
+  # ActiveRecord::ConnnectionAdapters::Transaction
+  # https://github.com/rails/rails/blob/4-2-stable/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb
+  class CallbackRecord
+    def initialize(&callback)
+      @callback = callback
+    end
 
-    def when_committed_callbacks
-      @when_committed_callbacks ||= []
+    def committed!
```

Comment:
The duck type of this object takes an optional parameter, only used [to notify records of failure in an ensure block](https://github.com/rails/rails/blob/ea11acd17a4c5dcbaa736a7caeee950c2a73c801/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb#L97).  You need to accept the param and decide what do do with it.